### PR TITLE
Fix catalog mode in products list and product page

### LINKED ISSUE #9937: [BOOM-3522] Native Catalog Mode Displays Product Prices
> This issue has been migrated from the Forge. Read the [original ticket here](http://forge.prestashop.com/browse/BOOM-3522).

- _**Reporter:**_ elpatron
- _**Created at:**_ Sat, 22 Jul 2017 20:02:26 +0200

<p>Hi ps team.  All previous versions of PrestaShop,  when native catalog mode enabled.  Would hide prices.</p>

<p>We noticed that 1.7 displays prices.  We suggest either hiding prices or better yet in conjunction with catalog mode have option to display or not display prices.</p>

<p>Loving 1.7, lots of new dev I could never do in previous versions, having template inheritance is a 'big deal'...hugs el</p>
        
* How to reproduce the issue ?

enable native prestashop catalog mode...see fo to see prices

### LINKED ISSUE #11742: Switch for price display when the catalog mode is on
**User Story**
As a merchant,  I want to choose if the prices are hidden or not when I enable the catalog mode.

**Description**
When I enable the catalog mode, it will display a switch to choose to show prices or not. By default, the prices are hidden so when the catalog mod is on the switch should be on "NO".

**Additional context**
mod catalogue

You can find the same UI comportment on Page Advanced Preferences > Performance > Smarty block > Enable/Disenable Cache

**Wording** 
In English:

**- Catalog mode**
_text on the bottom of the "Catalog mode" switch_:
> Disable products checkout (prices, add to cart buttons, price filters etc.)

Catalog mode toolbox: "Catalog mode makes your store look like a retail website."

**- Show prices**
_text on the bottom of the "Show prices" switch_ : None

Show prices toolbox: None

**Reference PR/Issues**
#9937

### LINKED ISSUE #12539: Wording clarification of "show prices" option
**Is your feature request related to a problem? Please describe.**
With the new catalog mode, Prestashop has 2 different options to hide the prices but for 2 different purposes. 
One on Product Setting is to hide all the prices in catalog mode as if your website is retail or a showcase and the other one Customer Setting to show the prices to a specific customer group like for only for the members.

This 2 options can be confusing for the new users, a new wordings should help them to understand more easily.

**Describe the solution you'd like**
The best option is to add a wording below the switch option "show prices".

Shop Parameters > Customer Setting > Edit/Add group > Show prices

> Need to hide prices for all groups? Save time, enable catalog mode in Product Settings instead.

Shop Parameters > Product Setting >Products (general) section > Catalog mode > Show prices

> Have specific needs? Edit particular groups to let them see prices or not. 

**Describe alternatives you've considered**
Wording changeable.

### PR DESCRIPTION
**Pull request by jolelievre**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | The catalog mode was not completely ported to 1.7 version, it was still showing prices in products listing and product page. The modification was managed in ProductLazyArray to avoid modifying any templates, this way any theme will benefit of this fix automatically. For merchants who wish to have a catalog mode with prices shown a new option was added in `Shop Parameters > Product settings`.
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Fixes #9937, fixes #11742 and fixes #12539
| How to test?  | Turn on "Catalog mode" in Shop Parameters -> Product Settings -> General Check that the prices are hidden in product listing. In product page the price and "add to cart" button are hidden. If you enable `Show prices` the prices are displayed also the cart and checkout buttons are still hidden.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/11688)

### CODE REVIEW

**Comment by MatShir** [comment]
Hi @jolelievre, 
Start some functional test, it seems that we can see the price filter on the categories list product. It can lead the customer to guess the price. More tests to come :)

**Comment by jolelievre** [comment]
@MatShir problem fixed. If you didn't find any other I ask for a code review.

**Comment by MatShir** [comment]
Hi @jolelievre,
Could you add on this PR, suppression of price sorting on the products list as well?

**Comment by jolelievre** [comment]
@MatShir done!
Waiting for review from @PrestaShop/prestashop-core-developers :)

File: classes/controller/ProductListingFrontController.php
Comment (PierreRambaud) [inline classes/controller/ProductListingFrontController.php]: Why `Configuration::get('PS_CATALOG_MODE') ` when you check for `!Configuration::get('PS_CATALOG_MODE') ` on line 399 :)

Comment (jolelievre) [reply classes/controller/ProductListingFrontController.php]: oups you're right, some old code I forgot to remove!

Comment (jolelievre) [reply classes/controller/ProductListingFrontController.php]: done!

File: src/Adapter/Presenter/Product/ProductLazyArray.php
Comment (PierreRambaud) [inline src/Adapter/Presenter/Product/ProductLazyArray.php]: Prefer `return !$this->settings->catalog_mode && $this->shouldShowPrice($this->settings, $this->product);`, it  will be faster in case `shouldShowPrice` has a complex behavior and `this->settings->catalog_mode` is already to false.

**Review by PierreRambaud — DISMISSED** [review]

File: src/Adapter/Presenter/Product/ProductLazyArray.php
Comment (eternoendless) [reply src/Adapter/Presenter/Product/ProductLazyArray.php]: Shouldn't this be done in `shouldShowPrice()` instead?

File: classes/controller/ProductListingFrontController.php
Comment (eternoendless) [inline classes/controller/ProductListingFrontController.php:399]: Please describe the purpose of this method in the Phpdoc header.

**Review by eternoendless — CHANGES_REQUESTED** [review]

File: classes/controller/ProductListingFrontController.php
Comment (jolelievre) [reply classes/controller/ProductListingFrontController.php:399]: described and renamed

File: admin-dev/themes/new-theme/js/pages/product-preferences/catalog-mode-option-handler.js
Comment (eternoendless) [inline admin-dev/themes/new-theme/js/pages/product-preferences/catalog-mode-option-handler.js]: All selectors must be extracted in a PageMap object in a separate file (to be used both by the page and E2E tests):

product-preferences-page-map.js:

```js
export default {
	catalogModeField: 'input[name="form[general][catalog_mode]',
	catalogModeOptions: '.catalog-mode-option',
	// ...
};
```

index.js
```js
import * as pageMap from 'product-preferences-page-map';

// ...
new CatalogModeOptionsHandler(pageMap);
```

File: classes/controller/ProductListingFrontController.php
Comment (eternoendless) [inline classes/controller/ProductListingFrontController.php]:
```suggestion
    protected function disablePriceControls(ProductSearchResult $result)
```

File: src/PrestaShopBundle/Resources/views/Admin/Configure/ShopParameters/Blocks/product_preferences_general.html.twig
Comment (eternoendless) [inline src/PrestaShopBundle/Resources/views/Admin/Configure/ShopParameters/Blocks/product_preferences_general.html.twig]: > Catalog mode makes your store look like a retail website.

Not sure about this wording. Isn't it supposed to disable checkout? How about this:

> Catalog mode disables the shopping cart on your store. Visitors will still able to browse your product catalog, but not buy them.

File: src/PrestaShopBundle/Resources/views/Admin/Configure/ShopParameters/Blocks/product_preferences_general.html.twig
Comment (eternoendless) [inline src/PrestaShopBundle/Resources/views/Admin/Configure/ShopParameters/Blocks/product_preferences_general.html.twig]: How about:

> Display product prices when in catalog mode.

File: classes/controller/ProductListingFrontController.php
Comment (eternoendless) [inline classes/controller/ProductListingFrontController.php]:
```suggestion
            $this->disabledPriceControls($result);
```

**Review by eternoendless — CHANGES_REQUESTED** [review]

File: src/PrestaShopBundle/Resources/views/Admin/Configure/ShopParameters/Blocks/product_preferences_general.html.twig
Comment (LouiseBonnard) [inline src/PrestaShopBundle/Resources/views/Admin/Configure/ShopParameters/Blocks/product_preferences_general.html.twig]: I think that there is a word missing: _Catalog mode disables the shopping cart on your store. Visitors will be able to browse your products catalog, but not buy them._

Comment (jolelievre) [reply src/PrestaShopBundle/Resources/views/Admin/Configure/ShopParameters/Blocks/product_preferences_general.html.twig]: True! fixed

**Review by eternoendless — CHANGES_REQUESTED** [review]
Looking good!

You should add an entry for the new setting in install-dev/data/xml/configuration.xml

> Reviewed 6 of 12 files at r5, 5 of 8 files at r6, 2 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/prestashop/prestashop/11688#-:-Lao_rw73vxBhl4cgZEk:bcj44uz)* status: all files reviewed, 4 unresolved discussions (waiting on @jolelievre, @LouiseBonnard, and @PierreRambaud)

___
*[admin-dev/themes/new-theme/js/pages/product-preferences/catalog-mode-option-handler.js, line 34 at r7](https://reviewable.io/reviews/prestashop/prestashop/11688#-LaoZKKy5T1UIV14qpq3:-LaoZKKz5WzAmFKMIt7_:bwydbex) ([raw file](https://github.com/prestashop/prestashop/blob/824694782d945eb304d7b3b16299b60d7c04907d/admin-dev/themes/new-theme/js/pages/product-preferences/catalog-mode-option-handler.js#L34)):*
> ```JavaScript
>       selectedCatalogModeField: 'input[name="form[general][catalog_mode]"]:checked',
>       catalogModeOptions: '.catalog-mode-option'
>     }, pageMap);
> ```

Why this `Object.assign`? You're effectively duplicating the page map.

**Comment by jolelievre** [comment]
___
*[admin-dev/themes/new-theme/js/pages/product-preferences/catalog-mode-option-handler.js, line 34 at r7](https://reviewable.io/reviews/prestashop/prestashop/11688#-LaoZKKy5T1UIV14qpq3:-LapEP8j-2YhIjlY9a_4:brr34r1) ([raw file](https://github.com/prestashop/prestashop/blob/824694782d945eb304d7b3b16299b60d7c04907d/admin-dev/themes/new-theme/js/pages/product-preferences/catalog-mode-option-handler.js#L34)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

Why this `Object.assign`? You're effectively duplicating the page map.
</blockquote></details>

I used this so that the component has default values. Besides I think it's easier to understand what the selector targets this way when you are reading since you don't need to go read another file.

**Review by jolelievre — COMMENTED** [review]
Oh I didn't know of this file! Thanks

> *[Reviewable](https://reviewable.io/reviews/prestashop/prestashop/11688#-:-LapErkjEFfZ80KVAhwP:b1ksrzb)* status: 13 of 14 files reviewed, 4 unresolved discussions (waiting on @eternoendless, @LouiseBonnard, and @PierreRambaud)

File: controllers/admin/AdminGroupsController.php
Comment (PierreRambaud) [inline controllers/admin/AdminGroupsController.php:370]: short array syntax :)

**Comment by jolelievre** [comment]
@LouiseBonnard just to be sure it's ok I added two help texts in:
- `controllers/admin/AdminGroupsController.php`
- `views/Admin/Configure/ShopParameters/Blocks/product_preferences_general.html.twig`

File: controllers/admin/AdminGroupsController.php
Comment (jolelievre) [reply controllers/admin/AdminGroupsController.php:370]: well all the file is in long array syntax.. it would not be consistent to use only one
And updating the whole file is not exactly in this PR scope

File: src/Adapter/Product/GeneralConfiguration.php
Comment (Quetzacoalt91) [inline src/Adapter/Product/GeneralConfiguration.php:74]: Isn't it weird for the merchant to have another value to be set? What would be the consequence to only rely on `$config['catalog_mode_with_prices']`?

**Review by Quetzacoalt91 — APPROVED** [review]
Only one question about the link between the two configuration switched. OK otherwise.

**Comment by Quetzacoalt91** [comment]
By the way @jolelievre, I did not find the upgrade content for the configuration value of `PS_CATALOG_MODE_WITH_PRICES` nor its default value in the Form.

File: src/Adapter/Product/GeneralConfiguration.php
Comment (jolelievre) [reply src/Adapter/Product/GeneralConfiguration.php:74]: the `PS_CATALOG_MODE_WITH_PRICES` is only taken into account when `PS_CATALOG_MODE` is on Otherwise it has no use (that's why the toggle button is hidden in the BO when catalog mode is off)

**Comment by jolelievre** [comment]
> By the way @jolelievre, I did not find the upgrade content for the configuration value of `PS_CATALOG_MODE_WITH_PRICES` nor its default value in the Form.

Well the default value is false so if the configuration doesn't exist in the database it will work anyway.
That said it's better to insert it in the upgrade script, which was managed in... (don't laugh) here:
https://github.com/PrestaShop/PrestaShop/pull/13048
😄

**Comment by sarahdib** [comment]
@jolelievre 
catalog mode set to « yes » and Show prices to « no »  
But I want to display the price of one product by set Show price to « yes » in the product page. it’s not working the price is still not display.

**Comment by jolelievre** [comment]
> @jolelievre
> catalog mode set to « yes » and Show prices to « no »
> But I want to display the price of one product by set Show price to « yes » in the product page. it’s not working the price is still not display.

@sarahdib I'm not sure but I think it is the desired behavior. When catalog mode prices are hidden (unless you activate "catalog with prices"). The "show product" option in the product page is used when you are not in a catalog mode (if you just want to hide some product's prices individually).

Anyway it is how it worked until today but this needs to be confirmed by @colinegin @samuel-pires @MatShir ?

**Review by eternoendless — APPROVED** [review]
> Reviewed 3 of 3 files at r8.
> *[Reviewable](https://reviewable.io/reviews/prestashop/prestashop/11688)* status: all files reviewed, 5 unresolved discussions (waiting on @LouiseBonnard, @PierreRambaud, and @Quetzacoalt91)

___
*[admin-dev/themes/new-theme/js/pages/product-preferences/catalog-mode-option-handler.js, line 34 at r7](https://reviewable.io/reviews/prestashop/prestashop/11688#-LaoZKKy5T1UIV14qpq3:-Lb3G-4GCdfPXPoA1HJz:bkyp88d) ([raw file](https://github.com/prestashop/prestashop/blob/824694782d945eb304d7b3b16299b60d7c04907d/admin-dev/themes/new-theme/js/pages/product-preferences/catalog-mode-option-handler.js#L34)):*
<details><summary><i>Previously, jolelievre wrote…</i></summary><blockquote>

I used this so that the component has default values. Besides I think it's easier to understand what the selector targets this way when you are reading since you don't need to go read another file.
</blockquote></details>

The principle of dependency injection is not to duplicate the dependency ;)

Not blocking but I think it should be removed

**Comment by jolelievre** [comment]
___
*[admin-dev/themes/new-theme/js/pages/product-preferences/catalog-mode-option-handler.js, line 34 at r7](https://reviewable.io/reviews/prestashop/prestashop/11688#-LaoZKKy5T1UIV14qpq3:-Lb3TtKX4-tYbdFKFtKP:b-gr1lld) ([raw file](https://github.com/prestashop/prestashop/blob/824694782d945eb304d7b3b16299b60d7c04907d/admin-dev/themes/new-theme/js/pages/product-preferences/catalog-mode-option-handler.js#L34)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

The principle of dependency injection is not to duplicate the dependency ;)

Not blocking but I think it should be removed
</blockquote></details>

I agree but injecting dependency doesn't mean you can't have default values

**Pull request merged**
